File: compiler/bench/minmax.fs
\ expected: 65536
\ Find min and max benchmark - 1M elements

1000000 constant SIZE
create arr SIZE cells allot

: arr@ ( i -- n ) cells arr + @ ;
: arr! ( n i -- ) cells arr + ! ;

: find-minmax ( -- min max )
  0 arr@ 0 arr@           \ min max
  SIZE 1 do
    i arr@
    dup 2 pick < if       \ new min
      rot drop swap
    else dup 2 pick > if  \ new max
      nip
    else
      drop
    then then
  loop ;

: init-arr ( -- )
  SIZE 0 do
    i 31337 * 65537 mod i arr!
  loop ;

: main
  init-arr
  find-minmax + . cr ;

File: compiler/bench/minmax.c
// expected: 65536
// Find min and max benchmark - 1M elements

#include <stdio.h>

#define SIZE 1000000

int arr[SIZE];

void find_minmax(int *min_out, int *max_out) {
    int min = arr[0], max = arr[0];
    for (int i = 1; i < SIZE; i++) {
        if (arr[i] < min) min = arr[i];
        else if (arr[i] > max) max = arr[i];
    }
    *min_out = min;
    *max_out = max;
}

void init_arr(void) {
    for (int i = 0; i < SIZE; i++) {
        arr[i] = ((long)i * 31337) % 65537;
    }
}

int main(void) {
    init_arr();
    int min, max;
    find_minmax(&min, &max);
    printf("%d\n", min + max);
    return 0;
}
